### PR TITLE
Add Feature Toggles - core classes

### PR DESCRIPTION
**Pull request by branimir-vujicic**
Feature Toggles should allow teams to modify system behavior without changing code. Feature Toggles are configured using google guice. Basic definition of toggles are crated using FeatureToggleBinder. FeatureToggleBinder creates FeatureToggle and additional configuration can be done using feature configuration.
In current stage Feature Toggles supports:
- if / else based feature toggles
- Dependency Injection based Hot reloading implementation without restart require code refactoring to add an interface when injecting the new implementation/class
- using various toggle strategies along with simple on / off toggles

configuration:

to allow feature toggle configuration four lines are needed in config.properties file
```
features.config-source-type=file
features.refresh-period=30s
```

`configuration-source-type` is source type for Feature Toggles configuration `features.config-source` is a source (file) of the configuration `features.config-type` format in which configuration is stored (json or properties) `features.refresh-period` configuration refresh period

Defining Feature Toggles

Feature toggle definition is done in google guice module using `FeatureToggleBinder`

simple feature toggle definition
```
    featureToggleBinder(binder)
                        .featureId("featureXX")
                        .bind()
```
This example creates bindings for @Inject
```
    @Inject
    public Runner(@FeatureToggle("featureXX") Supplier<Boolean> isFeatureXXEnabled)
    {
        this.isFeatureXXEnabled = isFeatureXXEnabled;
    }
```
`isFeatureXXEnabled` can be used to test if feature is enabled or disabled:
```
    boolean testFeatureXXEnabled()
    {
        return isFeatureXXEnabled.get();
    }
```

hot reloadable feature toggle definition
```
featureToggleBinder(binder, Feature01.class)
                        .featureId("feature01")
                        .baseClass(Feature01.class)
                        .defaultClass(Feature01Impl01.class)
                        .allOf(Feature01Impl01.class, Feature01Impl02.class)
                        .bind()
```
adding Feature Toggle switching strategy
```
featureToggleBinder(binder)
                        .featureId("feature04")
                        .toggleStrategy("AllowAll")
                        .toggleStrategyConfig(ImmutableMap.of("key", "value", "key2", "value2"))
```

feature-config.properties file example
```
# feature query-logger
feature.query-logger.enabled=true
feature.query-logger.strategy=OsToggle
feature.query-logger.strategy.os_name=.*Linux.*

#feature.query-rate-limiter
feature.query-rate-limiter.currentInstance=com.facebook.presto.server.protocol.QueryBlockingRateLimiter

# feature.query-cancel
feature.query-cancel.strategy=AllowList
feature.query-cancel.strategy.allow-list-source=.*IDEA.*
feature.query-cancel.strategy.allow-list-user=.*prestodb
```
in this example for first feature `query-logger`
changing value of feature.query-logger.enabled to `false` will 'disable' this feature. Changes will be effective within refresh period.

Test plan - Unit Tests
```
== RELEASE NOTES ==

General Changes
 * Add support for Feature Toggles.

### CODE REVIEW

**Comment by ajaygeorge** [comment]
@branimir-vujicic  I'm starting to take a look at the PR.
 I would recommend further breaking this up into commits like
1. Add Feature Toggle core classes
2. Add support for Simple feature toggle
3. Add support for Hot Reloadable feature toggle.
4. Add support for Strategy based feature toggle.
5. Add Test support for Feature toggle.

**Comment by branimir-vujicic** [comment]
@ajaygeorge, I'll try to break it down furthermore, and let you know.

File: presto-feature-toggle/README.md [hunk 1]
```diff
@@ -0,0 +1,233 @@
+# Feature Toggles
+
+Feature Toggles should allow teams to modify system behavior without changing code. Feature Toggles are configured using Google Guice. The basic definition of toggles is created
```
Comment (steveburnett) [inline presto-feature-toggle/README.md:3]:
```suggestion
Feature Toggles should allow teams to modify system behavior without changing code. Feature Toggles are configured using [Google Guice](https://github.com/google/guice). The basic definition of toggles is created
```

File: presto-feature-toggle/README.md [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+    1. [Simple feature toggle definition](#simple-feature-toggle-definition)
+    2. [Hot reloadable feature toggle definition](#hot-reloadable-feature-toggle-definition)
+    3. [Strategy-based Feature Toggling](#strategy-based-feature-toggling)
+3. [Examples](#examples)
+    1. [Query Cancel Feature](#query-cancel-feature)
+    2. [feature-config.properties file example](#feature-toggle-config-file-example)
+
+## Configuration
+
+To allow feature toggle configuration four lines are needed in the config.properties file
```
Comment (steveburnett) [inline presto-feature-toggle/README.md:19]:
```suggestion
To allow feature toggle configuration, add four lines in the config.properties file:
```
Suggest active voice

File: presto-feature-toggle/README.md [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+```
+
+- `configuration-source-type` is the source type for Feature Toggles configuration
+- `features.config-source` is a source (file) of the configuration
+- `features.config-type` format in which configuration is stored (JSON or properties)
+- `features.refresh-period` configuration refresh period
+
+## Defining Feature Toggles
+
+Feature toggle definition is done in the Google guice module using `FeatureToggleBinder`
```
Comment (steveburnett) [inline presto-feature-toggle/README.md:35]:
```suggestion
Feature toggle definition is done in the Google Guice module using `FeatureToggleBinder`.
```

File: presto-feature-toggle/README.md [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+Supplier&lt;Boolean&gt; `isFeatureXXEnabled` can be used to test if the feature is enabled or disabled:
+
+```
+    boolean testFeatureXXEnabled()
+    {
+      return isFeatureXXEnabled.get();
+    }
+```
+
+Switching the feature toggle on/off is done by changing the enabled value from true to false in the configuration source file:
```
Comment (steveburnett) [inline presto-feature-toggle/README.md:64]:
```suggestion
Switching the feature toggle on or off is done by changing the enabled value from `true` to `false` in the configuration source file:
```
Avoid slashes, see the [Gitlab documentation style guide entry for slashes](https://docs.gitlab.com/ee/development/documentation/styleguide/word_list.html#slashes).

File: presto-feature-toggle/README.md [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+
+```
+    feature.featureXX.enabled=true
+```
+
+After the refresh period value of `isFeatureXXEnabled.get();`  is changed.
```
Comment (steveburnett) [inline presto-feature-toggle/README.md:70]: This reads a little confusingly for me. Can you rephrase this?

File: presto-feature-toggle/README.md [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+        public String testFeature01()
+        {
+            return feature01.get().test();
+        }
+    }
+```
+
+## Strategy-based Feature Toggling
+
+Strategy-based Toggling allows us to Implement custom predicates (Strategy Pattern) to evaluate if a feature is enabled.
```
Comment (steveburnett) [inline presto-feature-toggle/README.md:103]:
```suggestion
Strategy-based Toggling allows us to implement custom predicates (Strategy Pattern) to evaluate if a feature is enabled.
```
Can you rephrase to remove the parentheses around "Strategy Pattern"? Also, if Strategy Pattern is an important concept here, consider adding a link to help people learn. Maybe the wikipedia entry is appropriate? 
https://en.wikipedia.org/wiki/Strategy_pattern

File: presto-feature-toggle/README.md [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+
+Some are provided out of the box: AllowAll, OS-based toggle, and AllowList toggle strategy.
```
Comment (steveburnett) [inline presto-feature-toggle/README.md:105]:
```suggestion
Default toggles provided are `AllowAll`, `AllowList`, and an operating system-based toggle strategy.
```

File: presto-feature-toggle/README.md [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+    }
+```
+
+## Feature Toggle config file example
+
+Any change of the configuration in the configuration file (source) overrides the feature toggle configuration.
+Allowed configuration params are:
+
+- enable
+- strategy.<param>
```
Comment (steveburnett) [inline presto-feature-toggle/README.md:216]: This line doesn't appear to display the way you seem to intend it to, see screenshot. 
Screenshot 2024-07-08 at 11 40 28 AM

File: presto-feature-toggle/README.md [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+
+```
+# feature.query-cancel
+feature.query-cancel.enable=true
+feature.query-cancel.strategy=AllowList
+feature.query-cancel.strategy.allow-list-source=.*IDEA.*
+feature.query-cancel.strategy.allow-list-user=.*prestodb
+```
+
+Configuration properties always start with a `feature` followed by a dot and the feature id.
```
Comment (steveburnett) [inline presto-feature-toggle/README.md:226]:
```suggestion
Configuration properties always start with a `feature` followed by a dot, then the feature id. In this example the feature id is `query-cancel`. 
```

File: presto-feature-toggle/README.md [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+
+Feature Toggle strategy properties start with `feature.featureId.strategy`. Property `feature.featureId.strategy` defines feature toggle strategy class (declared by registered
+name).
+After that, we can declare key-value pairs for parameters allowed for a given strategy.
+
+In this example for feature `query-cancel`, changing the value of feature.query-cancel.enabled to `false` will 'disable' this feature.
```
Comment (steveburnett) [inline presto-feature-toggle/README.md:232]:
```suggestion
In this example, change the value of `feature.query-cancel.enabled` to `false` to disable this feature.
```

File: presto-feature-toggle/README.md [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+Changes will be effective within the refresh period.
```
Comment (steveburnett) [inline presto-feature-toggle/README.md:233]:
```suggestion
Changes are effective within the refresh period. 
```
Does "within" mean the change is effective during the current refresh period, or after the refresh period?

**Review by steveburnett — CHANGES_REQUESTED** [review]
Thanks for the doc! Some format suggestions, and a couple of questions for you to consider and maybe change the text.

**Comment by branimir-vujicic** [comment]
Thank you, @steveburnett, for your suggestions. 
I have updated the document. Could you please review the changes?

**Review by steveburnett — DISMISSED** [review]
LGTM! (docs) 

Thanks for the revision! Used "View file" on the README to review the updated file as it will be displayed, verified all changes have been addressed, found no new comments to make. Nice work!
